src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs

```diff
@@ -89,5 +90,16 @@ private static IEnumerable<AbstractFormattingRule> GetFormattingRules(Document d
                 return snapshot.GetOpenDocumentInCurrentContextWithChanges();
             }
         }
+
+        /// <summary>
+        /// Get <see cref="Document"/> from <see cref="Text.Extensions.GetOpenDocumentInCurrentContextWithChanges(ITextSnapshot)"/>
+        /// once <see cref="IWorkspaceStatusService.WaitUntilFullyLoadedAsync(CancellationToken)"/> returns
+        /// </summary>
+        public static Document GetFullyLoadedOpenDocumentInCurrentContextWithChanges(
+            this ITextSnapshot snapshot, IUIThreadOperationContext operationContext, IThreadingContext threadingContext)
+        {
+            return threadingContext.JoinableTaskFactory.Run(() =>
```

Comment:
Just like we have WaitAndGetResult() assert if you're on a foreground thread to make sure nobody is accidentally blocking up backgrounds threads when they could have used the async form, does this need the same?

Reply:
:memo: This comment will likely be obsolete once https://github.com/dotnet/roslyn/pull/36864#discussion_r299219474 is addressed.

Reply:
I wrote big comment on why I like my approach. but github seems not showing the comments...

basically, I think the way it is done is safer than forcing everything to call JTF.Run(...)